planner/cascades/transformation_rules.go

```diff
@@ -2149,3 +2154,95 @@ func (r *InjectProjectionBelowTopN) OnTransform(old *memo.ExprIter) (newExprs []
 	topProjGroupExpr.SetChildren(topNGroup)
 	return []*memo.GroupExpr{topProjGroupExpr}, true, false, nil
 }
+
+// TransformApplyToJoin transforms a LogicalApply to LogicalJoin if it's
+// inner children has no correlated columns from it's outer schema.
+type TransformApplyToJoin struct {
+	baseRule
+}
+
+// NewRuleTransformApplyToJoin creates a new Transformation TransformApplyToJoin.
+// The pattern of this rule is: `Apply -> (X, Y)`.
+func NewRuleTransformApplyToJoin() Transformation {
+	rule := &TransformApplyToJoin{}
+	rule.pattern = memo.NewPattern(memo.OperandApply, memo.EngineTiDBOnly)
+	return rule
+}
+
+// OnTransform implements Transformation interface.
+func (r *TransformApplyToJoin) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	apply := old.GetExpr().ExprNode.(*plannercore.LogicalApply)
+	groupExpr := old.GetExpr()
+	// It's safe to use the old apply instead of creating a new LogicalApply here,
+	// Because apply.CorCols will only be used and updated by this rule during Transformation.
+	apply.CorCols = r.extractCorColumnsBySchema(groupExpr.Children[1], groupExpr.Children[0].Prop.Schema)
+	if len(apply.CorCols) != 0 {
+		return nil, false, false, nil
+	}
+
+	join := apply.LogicalJoin.Clone()
+	joinGroupExpr := memo.NewGroupExpr(join)
+	joinGroupExpr.SetChildren(groupExpr.Children...)
+	return []*memo.GroupExpr{joinGroupExpr}, true, false, nil
+}
+
+func (r *TransformApplyToJoin) extractCorColumnsBySchema(innerGroup *memo.Group, outerSchema *expression.Schema) []*expression.CorrelatedColumn {
+	corCols := r.extractCorColumnsFromGroup(innerGroup)
+	return plannercore.ExtractCorColumnsBySchema(corCols, outerSchema)
+}
+
+func (r *TransformApplyToJoin) extractCorColumnsFromGroup(g *memo.Group) []*expression.CorrelatedColumn {
+	corCols := make([]*expression.CorrelatedColumn, 0)
+	for elem := g.Equivalents.Front(); elem != nil; elem = elem.Next() {
+		expr := elem.Value.(*memo.GroupExpr)
+		corCols = append(corCols, expr.ExprNode.ExtractCorrelatedCols()...)
+		for _, child := range expr.Children {
+			corCols = append(corCols, r.extractCorColumnsFromGroup(child)...)
+		}
+	}
+	return corCols
```

Comment:
If there are duplicated correlated columns here, what will happen?

Reply:
See the implementation in `NestedLoopApplyExec.Next()`:
```golang
for _, col := range e.outerSchema {
    *col.Data = e.outerRow.GetDatum(col.Index, col.RetType)
}
```
So It's ok to have duplicated correlated columns. But it would be better to deduplicate the columns which could improve a litter perfomance. I prefer to do this in `ResolveIndices`.

Reply:
So we do it in `ResolveIndices` and add a comment here?